docker/parse.go

```diff
@@ -0,0 +1,111 @@
+// Parse a dockerfile into a high-level representation using the official go parser
+package docker
+
+import (
+	"io"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/moby/buildkit/frontend/dockerfile/command"
+	"github.com/moby/buildkit/frontend/dockerfile/parser"
+)
+
+// Represents a single line (layer) in a Dockerfile.
+// For example `FROM ubuntu:xenial`
+type Command struct {
+	Cmd       string   // lower cased command name (ex: `from`)
+	SubCmd    string   // for ONBUILD only this holds the sub-command
+	Json      bool     // whether the value is written in json form
+	Original  string   // The original source line
+	StartLine int      // The original source line number which starts this command
+	EndLine   int      // The original source line number which ends this command
+	Flags     []string // Any flags such as `--from=...` for `COPY`.
+	Value     []string // The contents of the command (ex: `ubuntu:xenial`)
+}
+
+// A failure in opening a file for reading.
+type IOError struct {
+	Msg string
+}
+
+func (e IOError) Error() string {
+	return e.Msg
+}
+
+// A failure in parsing the file as a dockerfile.
+type ParseError struct {
+	Msg string
+}
+
+func (e ParseError) Error() string {
+	return e.Msg
+}
+
+// List all legal cmds in a dockerfile
+func AllCmds() []string {
+	var ret []string
+	for k := range command.Commands {
+		ret = append(ret, k)
+	}
+	sort.Strings(ret)
+	return ret
+}
+
+// Parse a Dockerfile from a reader.  A ParseError may occur.
+func ParseReader(file io.Reader) ([]Command, error) {
+	res, err := parser.Parse(file)
+	if err != nil {
+		return nil, ParseError{err.Error()}
+	}
+
+	var ret []Command
+	for _, child := range res.AST.Children {
+		cmd := Command{
+			Cmd:       child.Value,
+			Original:  child.Original,
+			StartLine: child.StartLine,
+			EndLine:   child.EndLine,
+			Flags:     child.Flags,
+		}
+
+		// Only happens for ONBUILD
+		if child.Next != nil && len(child.Next.Children) > 0 {
+			cmd.SubCmd = child.Next.Children[0].Value
+			child = child.Next.Children[0]
+		}
+
+		cmd.Json = child.Attributes["json"]
+		for n := child.Next; n != nil; n = n.Next {
+			cmd.Value = append(cmd.Value, n.Value)
+		}
+
+		ret = append(ret, cmd)
+	}
+	return ret, nil
+}
+
+// Parse a Dockerfile from a filename.  An IOError or ParseError may occur.
+func ParseFile(filename string) ([]Command, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, IOError{err.Error()}
+	}
+	defer file.Close()
+
+	return ParseReader(file)
+}
+
+// Parse tag from parsed dockerfile:
+// e.g. FROM ubuntu:xenial returns "ubuntu", "xenial"
+func GetImageTagFromParsedFile(cmds []Command) (baseImage, tag string) {
```

Comment:
So Dockerfiles can have multiple "targets" in an image -- how does this cope with such a file? For example check out 

https://github.com/astronomer/ap-airflow/blob/7ffb41911ae106b9e110b8bb0e276a089ac1162b/1.10.5/buster/Dockerfile#L18
https://github.com/astronomer/ap-airflow/blob/7ffb41911ae106b9e110b8bb0e276a089ac1162b/1.10.5/buster/Dockerfile#L90
https://github.com/astronomer/ap-airflow/blob/7ffb41911ae106b9e110b8bb0e276a089ac1162b/1.10.5/buster/Dockerfile#L134

Reply:
@ashb i don't think we are going to parse something complicated, the idea is to check and warn that you are using custom dockerfile and we are not guarantee if it works.

FYI we can parse more complicated Dockerfiles:
```
(dlv) p cmds
[]github.com/astronomer/astro-cli/docker.Command len: 2, cap: 2, [
        {
                Cmd: "from",
                SubCmd: "",
                Json: false,
                Original: "FROM astronomerinc/ap-airflow:latest-onbuild as airflow-apt-deps",
                StartLine: 1,
                EndLine: 1,
                Flags: []string len: 0, cap: 0, [],
                Value: []string len: 3, cap: 4, [
                        "astronomerinc/ap-airflow:latest-onbuild",
                        "as",
                        "airflow-apt-deps",
                ],},
        {
                Cmd: "from",
                SubCmd: "",
                Json: false,
                Original: "FROM second:latest as main",
                StartLine: 2,
                EndLine: 2,
                Flags: []string len: 0, cap: 0, [],
                Value: []string len: 3, cap: 4, [
                        "second:latest",
                        "as",
                        "main",
                ],},
]
(dlv)
```

Reply:
To handle this "properly" we'd have to follow the chain "backwards" from the last `FROM` in the file.

I suspect this is so rarely used, especially in one of our projects that we can not worry about it for now.